sqlcollaborative/dbatools
# Creation of Devcontainer to support VS Code and GitHub Codespaces

Does not pertain to the direct codebase of the module. 

This is the addition of [DevContainers](https://code.visualstudio.com/docs/remote/create-dev-container) support for the repository for VS Code (remote container development) and [GitHub Codespaces](https://github.com/features/codespaces).

## VS Code Remote Container Development

To test for VS Code:
1. Clone repository
1. Do a `git fetch`
1. Checkout the devcontainer branch (`git checkout devcontainer`)
1. Reload VS Code
1. You should be prompted to reopen in a container

![image](https://user-images.githubusercontent.com/11204251/100529590-95123100-31ae-11eb-8ffb-e3d4527b1aa5.png)

I've added environment variables with the "servers" listed so you can easily grab what each container name is:
![image](https://user-images.githubusercontent.com/11204251/100529595-a8bd9780-31ae-11eb-83c1-aa6ebbb206c0.png)

```powershell
PS /workspace> $servers = $env:MY_SERVERS.Split(',')
PS /workspace> $servers
dbatools1
dbatools2
dbatools3
PS /workspace> 
```

After the above is done you can then proceed to create connections to each server based on the `SA_PASSWORD` that is set (since these are SQL Server containers at the base image):

![image](https://user-images.githubusercontent.com/11204251/100529659-66e12100-31af-11eb-826e-130195db7680.png)

```powershell
PS /workspace> Import-Module ./dbatools.psd1
PS /workspace> $cred = [pscredential]::new('sa',(ConvertTo-SecureString $env:SA_PASSWORD -AsPlainText -Force))
PS /workspace> $serversCn = Connect-DbaInstance -SqlInstance $servers -SqlCredential $cred
PS /workspace> $serversCn.NetName
ef4266d0e0a7
5dae5ae59158
519bccb6e0fc
PS /workspace> $serverCn | ft

ComputerName Name      Product              Version   HostPlatform IsAzure IsClustered ConnectedAs
------------ ----      -------              -------   ------------ ------- ----------- -----------
dbatools1    dbatools1 Microsoft SQL Server 15.0.4073 Linux        False   False       sa
dbatools2    dbatools2 Microsoft SQL Server 15.0.4073 Linux        False   False       sa
dbatools3    dbatools3 Microsoft SQL Server 15.0.4073 Linux        False   False       sa

PS /workspace> 
```

Being that VS Code reopens the workspace into the container you can either just close the VS Code window or utilize the command palette option to reopen in local:

![image](https://user-images.githubusercontent.com/11204251/100529812-a8260080-31b0-11eb-97aa-982ab03d033e.png)


## GitHub Codespaces

> Note this is in beta, so be aware miles may vary on experience

To test for GitHub Codespaces (requires your account has been added to the beta):
1. Navigate to the repo on GitHub
1. Switch to the devcontainer branch
1. Click the drop-down for Code
1. Click `Open with Codespaces`
1. Click `+ New codespace`

When you first create the codespace it will go through to initialize the environment for Codespaces itself, and then begin building the container and starting it.

![image](https://user-images.githubusercontent.com/11204251/100529870-2d111a00-31b1-11eb-9f4a-f87d2ea302e4.png)

If all goes well VS Code (web) will open and the workspace loaded:

![image](https://user-images.githubusercontent.com/11204251/100529923-b6285100-31b1-11eb-87e0-07026dc19d14.png)

> If you utilize VS Code sync feature and tied it to your GitHub user all your preferences and extensions will be installed; in addition to the ones set to install by default by the devcontainer definition.
